placehero/src/avatars.rs

```diff
@@ -23,79 +26,114 @@ struct AvatarResponse {
     image: Image,
 }
 
-#[derive(Default)]
+#[derive(Default, Debug)]
 pub(crate) struct Avatars {
     icons: HashMap<String, Option<Image>>,
     requester: Option<UnboundedSender<AvatarRequest>>,
     // TODO: lru: ...
 }
 
+impl Resource for Avatars {}
+
 impl Avatars {
-    pub(crate) fn worker(&mut self) -> impl View<Self, (), ViewCtx, Element = NoElement> + use<> {
-        worker(
-            |proxy: MessageProxy<AvatarResponse>, mut rx: UnboundedReceiver<AvatarRequest>| async move {
-                while let Some(url) = rx.recv().await {
-                    let proxy = proxy.clone();
-                    tokio::task::spawn(async move {
-                        let url = url.avatar_url;
-                        let result = image_from_url(&url).await;
-                        match result {
-                            // We choose not to handle the case where the event loop has ended
-                            Ok(image) => drop(proxy.message(AvatarResponse { url, image })),
-                            // TODO: Report in the frontend?
-                            Err(err) => {
-                                tracing::warn!("Loading avatar from {url:?} failed: {err}.");
-                            }
+    pub(crate) fn provide<State, Action, Child>(
+        child: Child,
+    ) -> impl WidgetView<State, Action, Element = Child::Element>
+    where
+        Child: WidgetView<State, Action>,
+        State: 'static,
+        Action: 'static,
+    {
+        provides(|_| Self::default(), fork(child, Self::worker()))
+    }
+
+    pub(crate) fn worker<State, Action>()
+    -> impl View<State, Action, ViewCtx, Element = NoElement> + use<State, Action>
+    where
+        State: 'static,
+        Action: 'static,
+    {
+        map_message(
+            on_action_with_context(
+                |_: &mut State, this: &mut Self, response| {
+                    let ret = this.icons.insert(response.url, Some(response.image));
+                    if !matches!(ret, Some(None)) {
+                        tracing::warn!("Potentially loaded or tried to load same avatar twice.");
+                    }
+                },
+                env_worker(
+                    |proxy: MessageProxy<AvatarResponse>,
+                     mut rx: UnboundedReceiver<AvatarRequest>| async move {
```

Comment:
(Not directly related to this PR)

I think a query based system similar to bevys systems could really help readability here (like in the callbacks of the `env_worker` maybe?)

And I think we should maybe look into refining the `OptionalAction` in `xilem_web` to maybe even `xilem_core` to avoid this `map_message`?